src/protocol/check_zero.rs

```diff
@@ -0,0 +1,149 @@
+
+    let r_times_v = ctx
+        .narrow(&Step::MultiplyWithR)
+        .multiply(record_id)
+        .await
+        .execute(r_sharing, v)
+        .await?;
+    let r_revealed = reveal(ctx.narrow(&Step::RevealR), record_id, r_times_v).await?;
+
+    Ok(r_revealed == F::ZERO)
+}
+
+#[cfg(test)]
+pub mod tests {
+    use crate::error::BoxError;
+    use crate::field::{Field, Fp31};
+    use crate::protocol::{check_zero::check_zero, QueryId, RecordId};
+    use crate::test_fixture::{logging, make_contexts, make_world, share, TestWorld};
+
+    #[tokio::test]
+    async fn basic() -> Result<(), BoxError> {
+        logging::setup();
+
+        let world: TestWorld = make_world(QueryId);
+        let context = make_contexts(&world);
+        let mut rng = rand::thread_rng();
+
+        for i in 0..u32::from(Fp31::PRIME) {
+            let a = Fp31::from(i);
+            let mut num_false_positives = 0;
+            for j in 0..10 {
+                let a_shares = share(a, &mut rng);
+                let row = i * u32::from(Fp31::PRIME) + j;
+                let record_id = RecordId::from(row);
+
+                let iteration = format!("{}", row);
+
+                let protocol_output = tokio::try_join!(
+                    check_zero(context[0].narrow(&iteration), record_id, a_shares[0],),
+                    check_zero(context[1].narrow(&iteration), record_id, a_shares[1],),
+                    check_zero(context[2].narrow(&iteration), record_id, a_shares[2],),
+                )?;
+
+                // All three helpers should always get the same result
+                assert_eq!(protocol_output.0, protocol_output.1);
+                assert_eq!(protocol_output.1, protocol_output.2);
+
+                match i {
+                    0 => {
+                        // When it actually is a secret sharing of zero
+                        // the helpers should definitely all receive "true"
+                        assert!(protocol_output.0);
+                        assert!(protocol_output.1);
+                        assert!(protocol_output.2);
+                    }
+                    _ => {
+                        // Unfortunately, there is a small chance of an incorrect
+                        // "true", even in when the secret shared value is NOT zero.
+                        // Since we will test out 10 different random secret sharings
+                        // let's count how many false positives we get. Odds are there
+                        // will be 0, 1, or maybe 2 out of 10
+                        if protocol_output.0 {
+                            num_false_positives += 1;
+                        }
+                    }
+                }
+            }
+
+            // Fp31 is just too small of a field.
+            // Through random chance, it'll incorrectly return "true"
+            // one time in 31. The odds of incorrectly returning "true"
+            // 5 times or more is... small...
```

Comment:
This one is correct, because you aren't counting when v==1 as a false positive :)